lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -136,6 +132,63 @@ import qualified Data.ByteString as BS
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as T
 
+-- @AddressScheme@ correlates @key@s with @Addresses@.
+--
+-- In a @SeqState@ we need tell if an @Address@ was created using a given @key@.
+--
+-- A @key@ may corrspond to multiple @Address@es. We cannot* convert
+-- an address into a public key.
+--
+-- To be able to compare @key@s and addresses, we convert them both to
+-- fingerprints, which are hashes of the public keys.
+--
+-- @@
+--
+--            paymentAddress         delegationAddress
+--
+--    Address ◀────────────── Key ──────────────▶ Address
+--        ╲                                         ╱
+--          ╲                  │                  ╱
+--            ╲                │                ╱
+--              ╲              │              ╱
+--                ╲            │            ╱
+--                  ╲          │          ╱
+--                    ╲        │        ╱
+--                      ╲      │      ╱
+--                        ╲    │    ╱
+--                          ╲  │  ╱
+--                            ╲│╱
+--                            ▼▼▼
+--                        Fingerprint
+-- @@
+--
+-- *) Except in jormungandr
+--
+-- TODO: @addressToText@ and @addressFromText@ could /maybe/ be removed if we made
+-- Address an anbstract type. But something entirely differnt could be desired
+-- as well.
+data AddressScheme (key :: Depth -> * -> *) = AddressScheme
+    { addressFromKey
+        :: key 'AddressK XPub -> Address
```

Comment:
This is missing the delegation addresses where addresses are constructed from two keys.

Reply:
> This is missing the delegation addresses where addresses are constructed from two keys.

I thought we could conflate `paymentAddress` and `delegationAddress` into one `addressFromKey`, where e.g. the `shelleyScheme` constructor takes the `Maybe StakingKey` as an argument, and produces `paymentAddress` if `Nothing`, and `delegationAddress` if `Just`.

https://github.com/input-output-hk/cardano-wallet/pull/1702/files/7a551ce9e87e717d3d0df85697db8a9a1d8ebedb#diff-708137f7529bf5efbae2de60278a966aR248-R263

Reply:
Hmmm. That would probably work fine for now indeed with the hypotheses under which the wallet is operating (i.e. one stake key per wallet and, using only delegated addresses).